junit-jupiter-api/src/main/java/org/junit/jupiter/api/AutoCloseExtension.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2015-2023 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api;
+
+import static org.junit.platform.commons.util.AnnotationUtils.findAnnotatedFields;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.function.Predicate;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
+import org.junit.jupiter.api.extension.TestInstancePreDestroyCallback;
+import org.junit.platform.commons.logging.Logger;
+import org.junit.platform.commons.logging.LoggerFactory;
+import org.junit.platform.commons.util.ExceptionUtils;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+/**
+ * {@code AutoCloseExtension} is a JUnit Jupiter extension that closes resources
+ * if a field in a test class is annotated with {@link AutoClose @AutoClose}.
+ *
+ * <p>Consult the Javadoc for {@link AutoClose @AutoClose} for details on the
+ * contract.
+ *
+ * @since 5.11
+ * @see AutoClose
+ */
+class AutoCloseExtension implements AfterAllCallback, TestInstancePreDestroyCallback {
+
+	private static final Logger logger = LoggerFactory.getLogger(AutoCloseExtension.class);
+	private static final Namespace NAMESPACE = Namespace.create(AutoClose.class);
+
+	@Override
+	public void afterAll(ExtensionContext context) {
+		Store contextStore = context.getStore(NAMESPACE);
+		Class<?> testClass = context.getRequiredTestClass();
+
+		registerCloseables(contextStore, testClass, null);
+	}
+
+	@Override
+	public void preDestroyTestInstance(ExtensionContext context) {
+		Store contextStore = context.getStore(NAMESPACE);
+
+		for (Object instance : context.getRequiredTestInstances().getAllInstances()) {
```

Comment:
I'm not sure we need to get all test instances here.

Rather, I think it should suffice to get only the current test instance.

In other words, if an enclosing class has any `@AutoClose` resources, those should be managed for the test instance of that enclosing class.

Or am I missing something?

Reply:
In order to ensure we have the correct semantics here, we'll need to introduce various `@Nested` test cases.

Reply:
For example, all tests in `AutoCloseTests` still pass if you change the method body to the following.

```java
		Store contextStore = context.getStore(NAMESPACE);
		Object testInstance = context.getRequiredTestInstance();
		registerCloseables(contextStore, testInstance.getClass(), testInstance);
```